libcudacxx/include/cuda/__mdspan/host_device_accessor.h

```diff
@@ -0,0 +1,383 @@
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+template <typename _Accessor>
+struct __host_accessor;
+
+template <typename _Accessor>
+struct __device_accessor;
+
+template <typename _Accessor>
+struct __managed_accessor;
+
+template <typename _Accessor>
+using host_accessor = __host_accessor<_Accessor>;
+
+template <typename _Accessor>
+using device_accessor = __device_accessor<_Accessor>;
+
+template <typename _Accessor>
+using managed_accessor = __managed_accessor<_Accessor>;
+
+/***********************************************************************************************************************
+ * Host/Device/Managed Accessor Traits
+ **********************************************************************************************************************/
+
+template <typename>
+inline constexpr bool is_host_accessor_v = false;
+
+template <typename>
+inline constexpr bool is_device_accessor_v = false;
+
+template <typename>
+inline constexpr bool is_managed_accessor_v = false;
+
+template <typename _Accessor>
+inline constexpr bool is_host_accessor_v<__host_accessor<_Accessor>> = true;
+
+template <typename _Accessor>
+inline constexpr bool is_device_accessor_v<__device_accessor<_Accessor>> = true;
+
+template <typename _Accessor>
+inline constexpr bool is_managed_accessor_v<__managed_accessor<_Accessor>> = true;
+
+template <typename _Tp>
+inline constexpr bool is_host_device_managed_accessor_v =
+  is_host_accessor_v<_Tp> || is_device_accessor_v<_Tp> || is_managed_accessor_v<_Tp>;
+
+/***********************************************************************************************************************
+ * Host Accessor
+ **********************************************************************************************************************/
+
+template <typename _Accessor>
+struct __host_accessor : public _Accessor
+{
+  static_assert(!is_host_device_managed_accessor_v<_Accessor>,
+                "cuda::__host_accessor/cuda::__device_accessor/cuda::__managed_accessor cannot be nested");
+
+  using offset_policy    = __host_accessor<typename _Accessor::offset_policy>;
+  using data_handle_type = typename _Accessor::data_handle_type;
+  using reference        = typename _Accessor::reference;
+  using element_type     = typename _Accessor::element_type;
+
+private:
+  static constexpr bool __is_access_noexcept = noexcept(_Accessor{}.access(data_handle_type{}, 0));
+  static constexpr bool __is_offset_noexcept = noexcept(_Accessor{}.offset(data_handle_type{}, 0));
+
+  template <typename data_handle_type>
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool __is_host_accessible_pointer(data_handle_type __p) noexcept
+  {
+    if constexpr (_CUDA_VSTD::is_pointer_v<data_handle_type>)
+    {
+      ::cudaPointerAttributes __attrib{};
+      _CCCL_VERIFY(::cudaPointerGetAttributes(&__attrib, __p) == ::cudaSuccess, "cudaPointerGetAttributes failed");
+      return __attrib.hostPointer != nullptr || __attrib.type == ::cudaMemoryTypeUnregistered;
+    }
```

Comment:
Important: We can extend this to `contiguous_iterator<data_handle_type>` and extract the pointer via `to_address` from `cuda/std/__memory/pointer_traits.h>`